### ISSUE TITLE
Applying a combination of both weighting and gradient-based methods can further improve performance？

### ISSUE THREAD

**Issue by puhan123** [lorenmt/auto-lambda#8]
As we know, weight-based methods search for different task weights, and task weights will act on the loss. Then the loss back-propagation will act on the gradient. In other words, weight-based methods have an effect on the gradient, and gradient-based methods also give different weights to each gradient. It looks like both weighting and gradient-based methods serve the same purpose.

I have three questions about the combination of both weighting and gradient-based methods :

1. What is the motivation for the combination of both weighting and gradient-based methods? 
2.  It looks like both weighting and gradient-based methods serve the same purpose. Why is it still necessary to combine both weighting and gradient-based methods? 
3. And why the combination of both can further improve performance?

Thank you very much.

**Comment by lorenmt** [comment]
Weighting-based methods simply find the best task weightings, and these weightings will be applied to the training loss and thus alter the task-specific gradients in an implicit way. However, these gradients can only be altered by changing their scale. And therefore, weighting-based methods typically are easy to optimise and run faster.

Gradient-based methods will directly alter task-specific gradients, and therefore give fine-grain control of how task-specific gradients should be like. Mathematically speaking, the range of gradients with weighting-based methods is a subset of one with gradient-based methods.

Empirically speaking, since the motivation and design logic for weighting and gradient-based methods are different, a better performance can be achieved by combining both methods (but it will run slower). You may consider that as we first use weighting-based methods to give a proper first guess, and then use gradient-based methods to refine the results.

**Comment by puhan123** [comment]
Thank you very much for your reply!

 According to your reply, I understand in this way: 

1. Weighting-based methods only alter the gradient scale, and Gradient-based methods will both alter the gradient direction and scale. 
2. The combination of both weighting and gradient-based methods is a coarse-to-fine process to alter the gradient.

 Otherwise, in your proposed Auto-lambda, if all tasks are considered primary tasks, the biggest difference with other weighting-based methods is that you take into account the validation loss， which will affect the model generalization.
Do I understand it properly?

**Comment by lorenmt** [comment]
> Weighting-based methods only alter the gradient scale, and Gradient-based methods will both alter the gradient direction and scale.
The combination of both weighting and gradient-based methods is a coarse-to-fine process to alter the gradient.

This is correct.

Auto-Lambda's comtributions are i) unify multi-task learning and auxiliary learning into a single framework -- as all previous methods need to have different designs for different different optimsiation problems. ii) Directly using the generlisation performance as a way to optimise task weightings. And yes, other weighting-based methods need to find an empirical measurement to determine which task is easy/hard. But we use a meta-formulation to mathematically parameterise "the generalisation performance", which (I think) is the better/closest definition of task difficulty.

**Comment by puhan123** [comment]
Thank you very much for your generous answer.

By using a meta-formulation to mathematically parameterize "the generalization performance",  in **the Design Philosophy** of the paper, you have denoted that  _The design of Auto-λ borrows the concept of looking ahead methods in meta-learning literature [8, 33], to update parameters at the current state of learning, based on the observed effect of those parameters on a future state.__


 "based on the observed effect of those parameters on a future state" means not only considering the current state of learning from the train data but also updating model parameters based on the learning effect in the validation data. The current state of learning means learning from the training data, and the future state of learning means learning from the validation data.

Do I understand this sentence correctly?

**Comment by lorenmt** [comment]
Yes.

But we only use validation data performance to update task weightings; training data performance is only to update network parameters.

**Comment by puhan123** [comment]
Thank you very much! Your answers have been very helpful to me.

**Comment by lorenmt** [comment]
You are welcome. And good luck with your research.
